### ISSUE TITLE
HABTM undefined method 'match'

### ISSUE THREAD

**Issue by willcosgrove** [rails/rails#23767]
I'm having an issue with a HABTM relationship that has a custom `class_name` set.  I poked around for a bit to try and figure out where the problem was, but I wasn't able to figure it out.  I am using Rails 5 beta 2, and I also tried out master, which also had the same problem.

I was able to find a workaround that worked for me.  I'll post the simplest setup that didn't work, and then my work around.

#### schema.rb
```ruby
  enable_extension "plpgsql"
  enable_extension "uuid-ossp"

  create_table "ballots", id: :uuid, default: -> { "uuid_generate_v4()" }, force: :cascade do |t|
    t.string   "name"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "ballots_users", id: false, force: :cascade do |t|
    t.uuid    "ballot_id"
    t.integer "user_id"
  end

  create_table "users", force: :cascade do |t|
    t.string   "email",            null: false
    t.datetime "created_at"
    t.datetime "updated_at"
  end
```

#### Models
```ruby
class User < ApplicationRecord
  has_and_belongs_to_many :invited_ballots, association_foreign_key: :ballot_id, class_name: Ballot
end

class Ballot < ApplicationRecord
  has_and_belongs_to_many :users
end
```

#### The error
_inside rails console_
```
[1] pry(main)> User.first.invited_ballots
Exception: NoMethodError: undefined method `match' for #<Class:0x007feac55c71c0>
Did you mean?  catch
--
 0: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/dynamic_matchers.rb:21:in `method_missing'
 1: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/inheritance.rb:137:in `compute_type'
 2: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/builder/has_and_belongs_to_many.rb:20:in `klass'
 3: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/builder/has_and_belongs_to_many.rb:14:in `join_table'
 4: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/builder/has_and_belongs_to_many.rb:57:in `table_name'
 5: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/core.rb:246:in `arel_table'
 6: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/core.rb:281:in `relation'
 7: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/scoping/named.rb:38:in `default_scoped'
 8: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/scoping/named.rb:28:in `all'
 9: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/through_association.rb:17:in `block in target_scope'
10: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/through_association.rb:16:in `each'
11: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/through_association.rb:16:in `target_scope'
12: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/association.rb:86:in `scope'
13: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/collection_association.rb:447:in `scope'
14: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/collection_proxy.rb:37:in `initialize'
15: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/relation/delegation.rb:100:in `new'
16: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/relation/delegation.rb:100:in `create'
17: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/collection_association.rb:47:in `reader'
18: /Users/will/.gem/ruby/2.3.0/bundler/gems/rails-454bc1deab3b/activerecord/lib/active_record/associations/builder/association.rb:111:in `invited_ballots'
```

#### The workaround

```ruby
class User < ApplicationRecord
  has_and_belongs_to_many :invited_ballots, association_foreign_key: :ballot_id, class_name: Ballot, join_table: :ballots_users
end
```

Let me know if there's any other info I can post to help debug this issue.  Thanks!

**Comment by mdoza** [comment]
@willcosgrove let Rails do all the heavy lifting for you, consider:

```ruby
class User < Application Record
  has_and_belongs_to_many :invited_ballots, class_name: "Ballot"
end

class Ballot < Application Record
  has_and_belongs_to_many :users
end
```

**Comment by willcosgrove** [comment]
@mdoza Interesting, this works
```ruby
class User < Application Record
  has_and_belongs_to_many :invited_ballots, class_name: "Ballot"
end
```
But this does not
```ruby
class User < Application Record
  has_and_belongs_to_many :invited_ballots, class_name: Ballot
end
```
I know the string'd way is the usual way to do it.  But providing the class constant works in all other cases, like a standard `belongs_to` or `has_many`.

**Comment by mdoza** [comment]
@willcosgrove I see what you're saying, it should be just as easy to pass the actual model class. I think that's why they key is explicitly named "class_name."

I agree it would be easy passing the model itself and have Rails convert to a string. But, what about models in different namespaces using that method?

**Comment by willcosgrove** [comment]
@mdoza Yeah that makes sense.  I used to use `class` rather than `class_name` which, confusingly, also worked, even though it was not part of the public api.  They then fixed that (#19659), so that you couldn't use `class`, so I switched to `class_name` but kept using constants because they still worked.

To your point on namespaced models, the table names are also namespaced by default.  So it should Just Work :tm:.

If there is a problem with using constants for the `class_name` setting, then they probably should never work, and just give an error letting the user know to use a string.  But I kinda like using constants to define the class name.  Feels nicer to me.
